tensorflow/compiler/xla/service/slice_sinker.cc

```diff
@@ -0,0 +1,276 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/slice_sinker.h"
+#include <algorithm>
+#include <utility>
+#include <vector>
+#include "absl/types/optional.h"
+#include "absl/container/flat_hash_set.h"
+#include "tensorflow/compiler/xla/shape_util.h"
+
+namespace xla {
+
+namespace {
+
+class SliceSinkerImpl {
+ public:
+  // Returns whether the instruction has been visited. The visited instruction
+  // has been grouped in a true user group with its peers and it should not be
+  // grouped in another true user group.
+  bool IsVisited(const HloInstruction* instruction) const;
+
+  // Collects true operands of inst. The inst's operands should all be slices,
+  // and the true operands are the operands of slices. Returns true operand
+  // vector if all true operands are found, or return an empty vector.
+  absl::optional<std::vector<HloInstruction*>>
+  FindSourceOperandsOfSlicesForElementwiseOperation(const HloInstruction* inst);
+
+  // Collects the true user group included inst. The slice from true operand is
+  // operand-slice. The true user's operands are all operand-slices.
+  // The operand-slices of a true user should be sliced from the same range of
+  // the true operands. It means the operand-slices have the same slice_starts,
+  // slice_limits, and slice_strides.
+  // The true users in a group have the same opcode and the same true operands
+  // in order. Of course, the inst is a true user, and the others in the group
+  // are the inst's peers.
+  // Returns an empty vector if the cost of change is more than the profit.
+  // Additionally, records the user which has been visited in a true user group,
+  // and skips it in later traverse.
+  absl::optional<std::vector<HloInstruction*>> FindPeerElementwiseOperations(
+      const HloInstruction* inst,
+      const std::vector<HloInstruction*>& operands);
+
+ private:
+  absl::flat_hash_set<const HloInstruction*> visited_;
+};
+
+// Computes the cost of implimentation of sinking slice, and returns whether
+// it should be changed.
+bool ShouldReplace(const std::vector<HloInstruction*>& operands,
+    const std::vector<HloInstruction*>& users) {
+  int64 sum = 0;
+  // Sums the total element number of the true users.
+  for (HloInstruction* user : users) {
+    sum += ShapeUtil::ElementsIn(user->shape());
+  }
+  // Operand and user have the same element number in shape because of
+  // elementwise operation, so the elements in operands[0] is the same as in
+  // new user if new user is generated.
+  // Compares the total elements in all true users and the elements of the new
+  // user with whole shape.
+  return sum >= xla::ShapeUtil::ElementsIn(operands[0]->shape());
+}
+
+// Generates the new user with the whole tensor and slices its output instead
```

Comment:
Generates a new elementwise operation using the slice_sources as operands, and replaces the uses of elementwise operation_on_slices with slices of the new elementwise operations.